SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,519 @@
+      if let p2ai = address as? PointerToAddressInst {
+        if let addrOrArg = pointerId.getOriginatingAddressOrArgument(p2ai) {
+          switch addrOrArg {
+          case let .address(newAddress):
+            return walkUp(address: newAddress, path: path)
+          case let .pointer(arg):
+            self.result = AccessPath(base: AccessBase(baseAddress: arg, kind: .pointer), projectionPath: path.projectionPath)
+          }
+        } else {
+          self.result = AccessPath(base: AccessBase(baseAddress: p2ai, kind: .pointer), projectionPath: path.projectionPath)
+          return .continueWalk
+        }
+      }
+
+      // If this is a base then we're done
+      if let base = AccessBase(baseAddress: address) {
+        self.result = AccessPath(base: base, projectionPath: path.projectionPath)
+        return .continueWalk
+      }
+
+      // The base is unidentified
+      self.result = nil
+      return .abortWalk
+    }
+
+    mutating func walkUp(address: Value, path: Path) -> WalkResult {
+      if address is IndexAddrInst {
+        // Track that we crossed an `index_addr` during the walk-up
+        return walkUpDefault(address: address, path: path.with(indexAddr: true))
+      } else if path.indexAddr && !canBeOperandOfIndexAddr(address) {
+        // An `index_addr` instruction cannot be derived from an address
+        // projection. Bail out
+        self.result = nil
+        return .abortWalk
+      } else if let ba = address as? BeginAccessInst, foundBeginAccess == nil {
+        foundBeginAccess = ba
+      }
+      return walkUpDefault(address: address, path: path.with(indexAddr: false))
+    }
+  }
+}
+
+/// A ValueUseDef walker that identifies which values a reference might
+/// originate from.
+protocol AccessStoragePathWalker : ValueUseDefWalker where Path == SmallProjectionPath {
+  mutating func visit(access: AccessStoragePath)
+}
+
+extension AccessStoragePathWalker {
+  // Main entry point
+  /// Given an `ap: AccessPath` where the base address is a projection
+  /// from a reference type, returns the set of reference definitions
+  /// that the address of the access may originate from.
+  mutating func getAccessStorage(for accessPath: AccessPath) {
+    walkUpCache.clear()
+    let path = accessPath.projectionPath
+    switch accessPath.base.baseAddress {
+    case let rea as RefElementAddrInst:
+      _ = walkUp(value: rea.operand, path: path.push(.classField, index: rea.fieldIndex))
+    case let pb as ProjectBoxInst:
+      _ = walkUp(value: pb.operand, path: path.push(.classField, index: pb.fieldIndex))
+    case let rta as RefTailAddrInst:
+      _ = walkUp(value: rta.operand, path: path.push(.tailElements, index: 0))
+    default:
+      visit(access: AccessStoragePath(storage: accessPath.base.baseAddress, path: accessPath.projectionPath))
+    }
+  }
+
+  // TODO: assert that argument is actually a reference (?)
+  mutating func getAccessStorage(for reference: Value) {
```

Comment:
I don't think we need this API function. When there is no `AccessPath` involved, `ValueUseDefWalker` can be used instead of `AccessStoragePathWalker`